Tools/clinic/clinic.py

```diff
@@ -73,6 +81,8 @@ def __repr__(self) -> str:
 NULL: Final = Sentinels.NULL
 unknown: Final = Sentinels.unknown
 
+NullType = type(Sentinels.NULL)
```

Comment:
Hmm, The type of `Sentinels.Null` is just `Sentinels`

Reply:
Oh, you're right.

Reply:
Perhaps we should amend this instead:

https://github.com/python/cpython/blob/dcdc90d384723920e8dea0ee04eae8c219333634/Tools/clinic/clinic.py#L2605-L2607

Reply:
I think my approach is fundamentally flawed. This is pretty simple really: we simply want distinct (sentinel) _types_ for `default_type` and the instance check (🥁).

Reply:
I'm not sure if we're in agreement or not (possibly my brain's going fuzzy at the end of a long day 😆), but I think for this PR, I might keep the refactor for `unspecified` and `unknown`, but leave `Null` as it is on `main`.

Reply:
Hm, I don't see any easy way to do this with enums. You want to give it a try?

Reply:
This also creates havoc:

```suggestion
NullType = typing.Type[Sentinels.NULL]
```

Reply:
> I'm not sure if we're in agreement or not (possibly my brain's going fuzzy at the end of a long day 😆), but I think for this PR, I might keep the refactor for `unspecified` and `unknown`, but leave `Null` as it is on `main`.

... or just leave it as it is for now and pick some low-hanging fruit instead 😄 I'm abandoning this PR for now.

Reply:
> This also creates havoc:

Well _that_ doesn't surprise me ;) `typing.Type[]` should only be parametrised by class objects, and `Sentinels.NULL` isn't a class object, it's an instance of the `Sentinels` enum ;)

 > ... or just leave it as it is for now and pick some low-hanging fruit instead 😄 I'm abandoning this PR for now.

Yeah, very fair! It looks like one or two of the previous annotations added might be incorrect, which might be complicating things here. I'm taking a look :)

Reply:
(Bear in mind I'm very new to `clinic.py`, still trying to understand this codebase somewhat):

What kinds of possible values do you expect `self.default` to have at line 3046 here? Could it be any object?

https://github.com/python/cpython/blob/dcdc90d384723920e8dea0ee04eae8c219333634/Tools/clinic/clinic.py#L3039-L3048

In the superclass, `default` is currently annotated as `bool | Unspecified`, but that looks a _little_ bit suspicious to me

Reply:
`self.default` is the Python default for the parameter in question; `bool` is definitely wrong.

Reply:
So it could be any object, right?

Reply:
`self.default` can literally be anything.

Reply:
Fab, thanks!

Reply:
The annotation for `default_type` is probably also incorrect.

https://github.com/python/cpython/blob/dcdc90d384723920e8dea0ee04eae8c219333634/Tools/clinic/clinic.py#L2598-L2607

Reply:
> The annotation for `default_type` is probably also incorrect.

It _looks_ correct to me based on the comment and usage. What makes you think it's incorrect?

Reply:
With a custom converter, you can override pretty much anything (AFAIK), so I _think_ `bltns.type[Any]` is too narrow. I might be wrong.

Reply:
The comment indicates that `default` has to be an instance of `default_type` if `default_type` is not `None`, and `isintance(x, Foo)` for any given `x` will fail if `Foo` is not an instance of type or a tuple of objects which are instances of `type`